ephios/core/services/matching.py

```diff
@@ -0,0 +1,139 @@
+        ]
+        return set(graph.spread_from(required_qualification_uuids))
+
+    @property
+    def skill_level(self):
+        return skill_level(self.required_skill)
+
+
+def skill_level(qualifications):
+    """
+    skill_level is some float in the range [0,1] that expresses
+    how far up in the qualification hierarchy the required qualifications are.
+    """
+    if not qualifications:
+        return 0
+    qualifications = list(qualifications)
+    # convert to uuids
+    if isinstance(qualifications[0], Qualification):
+        qualifications = [q.uuid for q in qualifications]
+    graph = QualificationUniverse.get_graph()
+    required_skill = set(graph.spread_from(qualifications))
+    # all_skill is qualifications up and down from the required ones
+    all_skill = required_skill | set(graph.spread_reverse(qualifications))
+    return len(required_skill) / len(all_skill)
+
+
+class Matching:
+
+    def __init__(self, participants, positions, pairings):
+        self.pairings = pairings
+        self.participants = participants
+        self.positions = positions
+        self.unpaired_participants = set(participants)
+        self.unpaired_positions = list(
+            sorted(positions, key=lambda position: (-position.skill_level, position.id))
+        )
+        for participant, position in pairings:
+            self.unpaired_participants.remove(participant)
+            self.unpaired_positions.remove(position)
+        self._participations = None
+
+    @property
+    def participations(self):
+        if self._participations is None:
+            raise ValueError("participations not attached")
+        return self._participations
+
+    def attach_participations(self, participations):
+        self._participations = participations
+
+    @property
+    def _participation_by_participant(self):
+        return {p.participant: p for p in self.participations}
+
+    @property
+    def participation_pairings(self):
+        return [
+            (self._participation_by_participant[participant], position)
+            for participant, position in self.pairings
+        ]
+
+    @property
+    def unpaired_participations(self):
+        return [
+            participation
+            for participation in self.participations
+            if participation.participant in self.unpaired_participants
+        ]
+
+
+def score_pairing(participant: AbstractParticipant, position: Position):
+    """
+    Score the pairing of participant and  position.
+    Skill here means a set of Qualification.
+    """
+    graph = QualificationUniverse.get_graph()
+    participant_skill = set(
+        graph.spread_from([qualification.uuid for qualification in participant.qualifications])
+    )
+    if not position.required_skill <= participant_skill:
+        # the participant does not have some required skill
+        return -1  #  avoid 0 as that causes trouble with the matching lib
```

Comment:
@BenBals thinks that it can occur that the algorithm will prefer a matching with someone unqualified because it guarantees that the matching fulfills  \lvert M \rvert = \min(\lvert U \rvert, \lvert V \rvert),

Reply:
Can he provide an example where it breaks? When we thought through it we came to the conclusion it shouldn't happen, no proof obvsl...